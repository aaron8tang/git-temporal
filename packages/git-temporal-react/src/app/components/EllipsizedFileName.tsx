import React from 'react';
import { style } from 'app/styles';

export interface EllipsizedFileNameProps {
  fileName: string;
  maxCharacters: number;
  style?: object | string;
}

export function ellipsizeFileName(
  fileName: string,
  maxCharacters: number
): string {
  const lengthIn = fileName.length;
  if (lengthIn > maxCharacters) {
    const overCount = lengthIn - maxCharacters + 4;
    const fileNameParts = fileName.split('/');
    let eliminatedCount = 0;
    let rightPath = '';
    for (let i = 2; i < fileNameParts.length - 1; i++) {
      const filePart = fileNameParts[i];
      if (eliminatedCount < overCount) {
        eliminatedCount += filePart.length + 1;
      } else {
        rightPath += `/${filePart}`;
      }
    }
    return `${fileNameParts[0]}/...${rightPath}/${
      fileNameParts[fileNameParts.length - 1]
    }`;
    // return `...${fileName.slice(lengthIn - maxCharacters + 3)}`;
  }
  return fileName;
}
export const EllipsizedFileName = (
  props: EllipsizedFileNameProps
): JSX.Element => {
  return (
    <span title={props.fileName} style={style(props.style)}>
      {ellipsizeFileName(props.fileName, props.maxCharacters)}
    </span>
  );
};
