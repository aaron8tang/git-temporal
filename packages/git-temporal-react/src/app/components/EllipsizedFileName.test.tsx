import React from 'react';
import { shallow } from 'enzyme';

import { EllipsizedFileName } from './EllipsizedFileName';

describe('components/EllipsizedFileName', () => {
  describe('when rendered with required props', () => {
    let wrapper;
    beforeAll(() => {
      wrapper = shallow(
        <EllipsizedFileName
          fileName="some/really/long/path/and/fileName.txt"
          maxCharacters={5}
        />
      );
    });

    test('it should have shortened the fileName without removing first and last parts)', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
