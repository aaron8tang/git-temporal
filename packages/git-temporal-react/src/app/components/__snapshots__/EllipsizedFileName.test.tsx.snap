// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/EllipsizedFileName when rendered with required props it should have shortened the fileName without removing first and last parts) 1`] = `
<span
  style={Object {}}
  title="some/really/long/path/and/fileName.txt"
>
  some/.../fileName.txt
</span>
`;
